src/analysis/normed_space/weak_dual_of_normed_space.lean

```diff
@@ -0,0 +1,146 @@
+variables {E : Type*} [normed_group E] [normed_space 𝕜 E]
+
+/-- For normed spaces `E`, there is a canonical map `dual 𝕜 E → weak_dual 𝕜 E` (the "identity"
+    mapping). It is a linear equivalence. -/
+def normed_space.dual.to_weak_dual : dual 𝕜 E ≃ₗ[𝕜] weak_dual 𝕜 E :=
+linear_equiv.refl 𝕜 (E →L[𝕜] 𝕜)
+
+/-- For normed spaces `E`, there is a canonical map `weak_dual 𝕜 E → dual 𝕜 E` (the "identity"
+    mapping). It is a linear equivalence. Here it is implemented as the inverse of the linear
+    equivalence `normed_space.dual.to_weak_dual` in the other direction. -/
+def weak_dual.to_original_dual : weak_dual 𝕜 E ≃ₗ[𝕜] dual 𝕜 E :=
+normed_space.dual.to_weak_dual.symm
+
+@[simp] lemma weak_dual.coe_to_fun_eq_original_coe_to_fun (x' : dual 𝕜 E) :
+  (x'.to_weak_dual : E → 𝕜) = x' := rfl
+
+@[simp] lemma to_weak_dual_inj_iff (x' y' : dual 𝕜 E) :
+  x'.to_weak_dual = y'.to_weak_dual ↔ x' = y' := iff.rfl
+
+@[simp] lemma to_original_dual_inj_iff (x' y' : weak_dual 𝕜 E) :
+  x'.to_original_dual = y'.to_original_dual ↔ x' = y' := iff.rfl
+
+theorem to_weak_dual_continuous :
+  continuous (λ (x' : dual 𝕜 E), x'.to_weak_dual) :=
+begin
+  apply weak_dual.continuous_of_continuous_eval,
+  intros z,
+  exact (inclusion_in_double_dual 𝕜 E z).continuous,
+end
+
+/-- For a normed space `E`, according to `to_weak_dual_continuous` the "identity mapping"
+`dual 𝕜 E → weak_dual 𝕜 E` is continuous. The following definition implements it as a continuous
+linear map. -/
+def normed_space.dual.continuous_linear_map_to_weak_dual : dual 𝕜 E →L[𝕜] weak_dual 𝕜 E :=
+{ cont := to_weak_dual_continuous,
+  .. normed_space.dual.to_weak_dual, }
+
+/-- The weak-star topology is coarser than the dual-norm topology: all weak-star open sets are open
+in the operator norm topology. -/
+lemma open_set_of_weak_dual_open_set (s : set (dual 𝕜 E))
+  (s_weak_dual_open : is_open (normed_space.dual.to_weak_dual '' s)) : is_open s :=
+begin
+  have eq : (normed_space.dual.to_weak_dual)⁻¹' (normed_space.dual.to_weak_dual '' s) = s,
+  { ext x',
+    simp only [set.mem_preimage, set.mem_image, to_weak_dual_inj_iff, exists_eq_right], },
+  rw ←eq,
+  apply continuous_def.mp to_weak_dual_continuous _ s_weak_dual_open,
+end
+
+-- TODO: The proof below may be abusing definitional equality... And it looks like it needs golf.
+private lemma to_weak_dual_image (s : set (dual 𝕜 E)) :
+  (normed_space.dual.to_weak_dual '' s) = s :=
+begin
+  ext x',
+  split,
+  { intros hx',
+    rcases hx' with ⟨y', hy', h_eq⟩,
+    rwa ←h_eq, },
+  { intros hx',
+    use x',
+    exact ⟨ hx', by refl ⟩, },
+end
+
+-- TODO: The proof and even the statement below may be abusing definitional equality...
+-- But I don't think this can be stated using `≤` on topologies without such abuse.
+/-- The weak-star topology is coarser than the dual-norm topology. -/
+theorem dual_norm_topology_le_weak_dual_topology :
+  (by apply_instance : topological_space (dual 𝕜 E)) ≤ weak_dual.weak_dual_topology 𝕜 E :=
+begin
```

Comment:
I initially I disliked the version that required identifying (via defeq) the `dual` and `weak_dual`, because I thought one should only move between these using the API maps `to_weak_dual` etc. That's why I wanted one formulation which explicitly used those maps.

I have now become less worried about the defeq abuse. In fact, would it make sense to even introduce coercions between the `weak_dual` and `dual`, so as to encourage thinking of them as the same set (as mathematicians would)? Is it a bad design idea to have a coercion both ways?

(I'll think about removing the `open_set_of_weak_dual_open_set`, if I get an efficient proof of `dual_norm_topology_le_weak_dual_topology` without it.)

Reply:
Ok, starting with your suggestion, the following 4 lines is what I get. 

```
theorem dual_norm_topology_le_weak_dual_topology :
  (by apply_instance : topological_space (dual 𝕜 E)) ≤ weak_dual.weak_dual_topology 𝕜 E :=
begin
  refine continuous.le_induced _,
  apply continuous_pi_iff.mpr,
  intros z,
  exact (inclusion_in_double_dual 𝕜 E z).continuous,
end
```

I'm sure you'd have done this more efficiently, but I think this is already enough to just get rid of `open_set_of_weak_dual_open_set`, which anyway is a transparent consequence.

Reply:
> I have now become less worried about the defeq abuse. In fact, would it make sense to even introduce coercions between the `weak_dual` and `dual`, so as to encourage thinking of them as the same set (as mathematicians would)? Is it a bad design idea to have a coercion both ways?

There is already a sort of coercion from the defeq-ness, as indeed you're seeing in this lemma.  If I understand correctly (not completely sure), writing in a formal coercion between definitionally-equal things doesn't really do anything ... when you write it, Lean happily tells you, "yes, I have checked and it makes sense to consider an element of the weak dual as an element of the normed dual", but it doesn't actually change anything.

In short I think your current version is at exactly the right level of defeq-abuse!